zproject/backends.py

```diff
@@ -2802,6 +2876,89 @@ def validate_otp_params(
+        this session is stored in authentication_method session attribute.
+        If SAML was used, this extracts the IdP name and returns it.
+        """
+        # Some asserts to ensure this doesn't get called incorrectly:
+        assert hasattr(request, "user")
+        assert isinstance(request.user, UserProfile)
+
+        authentication_method = request.session.get("authentication_method", "")
+        if not authentication_method.startswith("saml:"):
+            return None
+
+        return authentication_method.split("saml:")[1]
+
+    @classmethod
+    def get_logged_in_user_session_index(cls, request: HttpRequest) -> Optional[str]:
+        """
+        During SAML authentication, we obtain the SessionIndex value provided
+        by the IdP and save it in the session. This function can be used
+        to retrieve it.
+        """
+        # Some asserts to ensure this doesn't get called incorrectly:
+        assert hasattr(request, "user")
+        assert isinstance(request.user, UserProfile)
+
+        session_index = request.session.get("saml_session_index")
+        return session_index
+
+    @classmethod
+    def slo_request_to_idp(
+        cls, request: HttpRequest, return_to: Optional[str] = None
+    ) -> HttpResponse:
+        """
+        Generates the redirect to the IdP's SLO endpoint with
+        the appropriately generated LogoutRequest or None if the session
+        wasn't authenticated via SAML.
+        """
+
+        user_profile = request.user
+        assert isinstance(user_profile, UserProfile)
+
+        realm = user_profile.realm
+        assert saml_auth_enabled(realm)
+
+        complete_url = reverse("social:complete", args=("saml",))
+        saml_backend = load_backend(load_strategy(request), "saml", complete_url)
+
+        idp_name = cls.get_logged_in_user_idp(request)
+        if idp_name is None:
+            raise AssertionError("User not logged in via SAML")
+        session_index = cls.get_logged_in_user_session_index(request)
+
+        idp = saml_backend.get_idp(idp_name)
+        auth = saml_backend._create_saml_auth(idp)
+        slo_url = auth.logout(
+            name_id=user_profile.delivery_email, return_to=return_to, session_index=session_index
+        )
+
+        return HttpResponseRedirect(slo_url)
+
+    @classmethod
+    def process_logout_response(cls, logout_response: SAMLResponse, idp_name: str) -> HttpResponse:
+        """
+        Validates the LogoutResponse and logs out the user if successful,
+        finishing the SP-initiated logout flow.
+        """
+        from django.contrib.auth.views import logout_then_login as django_logout_then_login
+
+        idp = logout_response.backend.get_idp(idp_name)
+        auth = logout_response.backend._create_saml_auth(idp)
+        auth.process_slo(keep_local_session=True)
+        errors = auth.get_errors()
+        if errors:
+            raise JsonableError(f"LogoutResponse error: {errors}")
```

Comment:
This sounds user-facing; do we have an example of what this might look like?  I wonder whether the `LogoutResponse error: {errors}` formatting is a good way to do it or not.

Reply:
This kinda user-facing, but errors here should essentially only happen in the case of misconfiguration, so making this mostly admin-facing. The error will be a json error page, so things like this:
![image](https://github.com/zulip/zulip/assets/45007152/d35cc44b-588c-4688-b20d-67a3dade4836)

So not a good error page for a user, but normal users shouldn't really be encountering them, and in this form it might be the most useful page for an admin testing this :thinking: These error come from inside `python3-saml`, which makes them pretty generic while being informative, in a way that is not going to leak some sensitive details:
```
            if not self.validate_response_signature(get_data):
                self._errors.append('invalid_logout_response_signature')
                self._errors.append('Signature validation failed. Logout Response rejected')
            elif not logout_response.is_valid(self._request_data, request_id):
                self._errors.append('invalid_logout_response')
                self._error_reason = logout_response.get_error()
            elif logout_response.get_status() != OneLogin_Saml2_Constants.STATUS_SUCCESS:
                self._errors.append('logout_not_success')
```

Reply:
I added a comment here to clarify that, since it deserves at least that clarification. Or do we want to show other kind of error page here?

Reply:
I think that's probably OK given the context in which this should appear.